InvenTree/plugin/installer.py

```diff
@@ -0,0 +1,223 @@
+            return
+
+    # Append plugin to file
+    lines.append(f'{install_name}')
+
+    # Write file back to disk
+    try:
+        with open(pf, 'w') as f:
+            for line in lines:
+                f.write(line)
+
+                if not line.endswith('\n'):
+                    f.write('\n')
+    except Exception as exc:
+        logger.error(f"Failed to add plugin to plugins file: {str(exc)}")
+
+
+def install_plugin(url, packagename=None, user=None):
+    """Install a plugin into the python virtual environment:
+
+    - A staff user account is required
+    - We must detect that we are running within a virtual environment
+    """
+
+    if user and not user.is_staff:
+        raise ValidationError(_("Permission denied: only staff users can install plugins"))
+
+    logger.info(f"install_plugin: {url}, {packagename}")
+
+    # Check if we are running in a virtual environment
+    # For now, just log a warning
+    in_venv = sys.prefix != sys.base_prefix
+
+    if not in_venv:
+        logger.warning("InvenTree is not running in a virtual environment")
+
+    # build up the command
+    install_name = []
+
+    if url:
+        # use custom registration / VCS
+        if True in [identifier in url for identifier in ['git+https', 'hg+https', 'svn+svn', ]]:
+            # using a VCS provider
+            if packagename:
+                install_name.append(f'{packagename}@{url}')
+            else:
+                install_name.append(url)
+        else:  # pragma: no cover
+            # using a custom package repositories
+            # This is only for pypa compliant directory services (all current are tested above)
+            # and not covered by tests.
+            install_name.append('-i')
+            install_name.append(url)
+            install_name.append(packagename)
+
+    elif packagename:
+        # use pypi
+        install_name.append(packagename)
+
+    ret = {}
+
+    # Execute installation via pip
+    try:
+        result = pip_command('install', *install_name)
+
+        ret['result'] = _("Installed plugin successfully")
+        ret['output'] = str(result, 'utf-8')
+
+        if path := check_package_path(packagename):
+            # Override result information
+            ret['result'] = _(f"Installed plugin into {path}")
+
+    except subprocess.CalledProcessError as error:
+        # If an error was thrown, we need to parse the output
+
+        output = error.output.decode('utf-8')
+        logger.error(f"Plugin installation failed: {output}")
+
+        errors = []
+
+        for msg in output.split("\n"):
+            msg = msg.strip()
+
+            if msg:
+                errors.append(msg)
+
+        if len(errors) > 1:
+            raise ValidationError(errors)
+        else:
+            raise ValidationError(errors[0])
+
+    # Save plugin to plugins file
+    add_plugin_to_file(' '.join(install_name))
+
+    # TODO: run migrations for this plugin
```

Comment:
Suggestion: There should be a setting for this - migrations can block the database for a long time and thus make the whole instance unusable suddently.

Reply:
One thing that makes installing an app plugin very "difficult" for the novice user is that the migrations don't get applied cleanly, and automatically.

I think that the approach is when a plugin (which supports the `app` mixin) is *activated* (not installed), then migrations are run then. It could be pushed off to a background worker too (which is the current implementation), but it currently does not work. Needs some investigation.

Reply:
A more complex AppMixin plugin will block the database 60s easily - that is something to keep in mind.

Reply:
Auto plugin migration and reload is now working quite well :)